src/main/java/com/spotify/autoscaler/Autoscaler.java

```diff
@@ -85,33 +88,40 @@ public AutoscaleJob makeAutoscaleJob(
 
   @Override
   public void run() {
-    /*
-     * Without this horrible bit of horribleness,
-     * any uncaught Exception would kill the whole autoscaler.
-     */
+    Future<?> runResult = executorService.submit(this::check);
```

Comment:
I prefer not having this outside the `try` as `submit` can throw an exception and basically kill the whole process. Maybe keeping this part as is, and adding the timeout to the `check` method instead would be better?
`CompletableFuture.allOf(futures).join();` can be `CompletableFuture.allOf(futures).get(TIMEOUT);` instead?

Reply:
I don't like adding the timeout at the end of `check` because there is a number of things in that method that are blocking (example: database calls). So I would wrap the entire call and timeout that, or timeout each individual thing that we know could fail/take long/misbehave.

Reply:
That sounds good, then maybe we should increase the thread pool size for the executor as we are now using it for the main thing as well, or use a separate executor for it? I don't know if that'd matter that much though, whatever you decide :)